penrose-web/src/compiler/Style.ts

```diff
@@ -1,7 +1,2068 @@
+  m.sTypeVarMap[toString(k)] = v;
+  m.varProgTypeMap[toString(k)] = [p, k];
+  return m;
+};
+
+// TODO: Test this
+// Judgment 3. G; g |- |S_o ok ~> g'
+// `checkDeclPattern`
+const checkDeclPatternAndMakeEnv = (varEnv: Env, selEnv: SelEnv, stmt: DeclPattern): SelEnv => {
+  const [styType, bVar] = [stmt.type, stmt.id];
+
+  const typeErr = checkTypeConstructor(toSubstanceType(styType), varEnv);
+  if (isErr(typeErr)) {
+    // TODO(errors)
+    throw Error("substance type error");
+  }
+
+  const varName: string = bVar.contents.value;
+
+  // TODO(errors)
+  if (Object.keys(selEnv.sTypeVarMap).includes(varName)) {
+    throw Error("Style pattern statement has already declared the variable");
+  }
+
+  if (bVar.tag === "StyVar") {
+    // rule Decl-Sty-Context
+    // NOTE: this does not aggregate *all* possible errors. May just return first error.
+    // y \not\in dom(g)
+    return addMapping(bVar, styType, selEnv, { tag: "StyProgT" });
+
+  } else if (bVar.tag === "SubVar") {
+    // rule Decl-Sub-Context
+    // x \not\in dom(g)
+
+    const substanceType = varEnv.vars.get(varName);
+    // If any Substance variable doesn't exist in env, ignore it,
+    // but flag it so we know to not translate the lines in the block later.
+    if (!substanceType) {
+      return { ...selEnv, skipBlock: true };
+    }
+
+    // check "T <: |T", assuming type constructors are nullary
+    // Specifically, the Style type for a Substance var needs to be more general. Otherwise, if it's more specific, that's a coercion
+    // e.g. this is correct: Substance: "SpecialVector `v`"; Style: "Vector `v`"
+    const declType = toSubstanceType(styType);
+    if (!isDeclaredSubtype(substanceType, declType, varEnv)) { // COMBAK: Order?
+      // TODO(errors)
+      throw Error("mismatched types or wrong subtypes b/t Sub and Sty vars");
+    }
+
+    return addMapping(bVar, styType, selEnv, { tag: "SubProgT" });
+
+  } else throw Error("unknown tag");
+};
+
+// Judgment 6. G; g |- [|S_o] ~> g'
+// `checkDeclPatterns` w/o error-checking, just addMapping for StyVars and SubVars
+const checkDeclPatternsAndMakeEnv = (varEnv: Env, selEnv: SelEnv, decls: DeclPattern[]): SelEnv => {
+  return decls.reduce((s, p) => checkDeclPatternAndMakeEnv(varEnv, s, p), selEnv);
+};
+
+// TODO: Test this function
+// Judgment 4. G |- |S_r ok
+const checkRelPattern = (varEnv: Env, rel: RelationPattern): StyErrors => {
+  // rule Bind-Context
+  if (rel.tag === "RelBind") {
+    // TODO: use checkSubStmt here (and in paper)?
+    // TODO: make sure the ill-typed bind selectors fail here (after Sub statics is fixed)
+    // G |- B : T1
+    const res1 = checkVar(rel.id.contents, varEnv);
+
+    // TODO (error)
+    if (isErr(res1)) {
+      // TODO(error): extract and return error -- how to convert from SubstanceError to StyError?
+      return ["substance typecheck error in B"];
+    }
+
+    const [vtype, env1] = unsafelyUnwrap(res1);
```

Comment:
The right way is to do this in the `else`. TS will infer the types if you call `res1.value`